File: docs/03-encodings/examples.py
#!/usr/bin/env python3
"""
Статья 3 — Кодировки: UTF-8, UTF-16, UTF-32.
Практические примеры кодирования/декодирования и байтового анализа.
"""


# ─────────────────────────────────────────────
# 1. Кодирование одного символа в разных UTF
# ─────────────────────────────────────────────

def show_encodings(char: str) -> None:
    """Показать байты символа во всех основных кодировках."""
    cp = ord(char)
    name = __import__('unicodedata').name(char, "?")
    print(f"Символ: {char!r}  U+{cp:05X}  {name}")
    for enc in ['utf-8', 'utf-16-le', 'utf-16-be', 'utf-32-le', 'utf-32-be']:
        try:
            b = char.encode(enc)
            hex_str = ' '.join(f'{byte:02X}' for byte in b)
            print(f"  {enc:15s}: [{hex_str}]  ({len(b)} байт)")
        except Exception as e:
            print(f"  {enc:15s}: ошибка — {e}")
    print()


print("=== Кодирование символов в разных UTF ===\n")
for ch in ['A', 'А', 'ñ', '中', '🌍', '\u0000']:
    show_encodings(ch)


# ─────────────────────────────────────────────
# 2. Ручное кодирование в UTF-8
# ─────────────────────────────────────────────

def encode_utf8_manual(cp: int) -> bytes:
    """Закодировать кодовую точку в UTF-8 вручную (без stdlib)."""
    if cp < 0 or cp > 0x10FFFF:
        raise ValueError(f"Недопустимая кодовая точка: {cp}")
    if cp <= 0x7F:
        return bytes([cp])
    elif cp <= 0x7FF:
        b1 = 0b11000000 | (cp >> 6)
        b2 = 0b10000000 | (cp & 0b00111111)
        return bytes([b1, b2])
    elif cp <= 0xFFFF:
        b1 = 0b11100000 | (cp >> 12)
        b2 = 0b10000000 | ((cp >> 6) & 0b00111111)
        b3 = 0b10000000 | (cp & 0b00111111)
        return bytes([b1, b2, b3])
    else:
        b1 = 0b11110000 | (cp >> 18)
        b2 = 0b10000000 | ((cp >> 12) & 0b00111111)
        b3 = 0b10000000 | ((cp >> 6) & 0b00111111)
        b4 = 0b10000000 | (cp & 0b00111111)
        return bytes([b1, b2, b3, b4])


def decode_utf8_manual(data: bytes) -> list[int]:
    """Декодировать UTF-8 байты в список кодовых точек (вручную)."""
    result = []
    i = 0
    while i < len(data):
        b = data[i]
        if b & 0b10000000 == 0:          # 1 байт
            cp = b
            i += 1
        elif b & 0b11100000 == 0b11000000:  # 2 байта
            cp = (b & 0b00011111) << 6
            cp |= data[i + 1] & 0b00111111
            i += 2
        elif b & 0b11110000 == 0b11100000:  # 3 байта
            cp = (b & 0b00001111) << 12
            cp |= (data[i + 1] & 0b00111111) << 6
            cp |= data[i + 2] & 0b00111111
            i += 3
        else:                               # 4 байта
            cp = (b & 0b00000111) << 18
            cp |= (data[i + 1] & 0b00111111) << 12
            cp |= (data[i + 2] & 0b00111111) << 6
            cp |= data[i + 3] & 0b00111111
            i += 4
        result.append(cp)
    return result


print("=== Ручное кодирование UTF-8 ===\n")
test_cps = [0x41, 0xC3, 0x410, 0x4E2D, 0x1F30D]
for cp in test_cps:
    ch = chr(cp)
    manual = encode_utf8_manual(cp)
    stdlib = ch.encode('utf-8')
    match = "✓" if manual == stdlib else "✗"
    print(f"U+{cp:05X}  {ch}  manual={manual.hex()}  stdlib={stdlib.hex()}  {match}")

print()
print("=== Ручное декодирование UTF-8 ===\n")
test_bytes = "Hello, Мир! 🌍".encode('utf-8')
decoded = decode_utf8_manual(test_bytes)
reconstructed = ''.join(chr(cp) for cp in decoded)
print(f"Исходная строка:  'Hello, Мир! 🌍'")
print(f"Восстановленная:  {reconstructed!r}")
print(f"Совпадает: {reconstructed == 'Hello, Мир! 🌍'}")
print()


# ─────────────────────────────────────────────
# 3. Суррогатные пары в UTF-16
# ─────────────────────────────────────────────

def encode_utf16_surrogate_pair(cp: int) -> tuple[int, int]:
    """Вычислить суррогатную пару для символа вне BMP."""
    assert 0x10000 <= cp <= 0x10FFFF, "Суррогаты только для SMP и выше"
    cp_prime = cp - 0x10000
    high = 0xD800 + (cp_prime >> 10)
    low  = 0xDC00 + (cp_prime & 0x3FF)
    return high, low


def decode_utf16_surrogate_pair(high: int, low: int) -> int:
    """Восстановить кодовую точку из суррогатной пары."""
    assert 0xD800 <= high <= 0xDBFF, "Не high surrogate"
    assert 0xDC00 <= low  <= 0xDFFF, "Не low surrogate"
    return 0x10000 + ((high - 0xD800) << 10) + (low - 0xDC00)


print("=== Суррогатные пары UTF-16 ===\n")
smp_chars = ['🌍', '𝄞', '𠀀', '🎉', '👨‍👩‍👧']
for ch in smp_chars:
    for sub_ch in ch:  # в случае emoji ZWJ, проходим каждый code point
        cp = ord(sub_ch)
        if cp > 0xFFFF:
            high, low = encode_utf16_surrogate_pair(cp)
            restored = decode_utf16_surrogate_pair(high, low)
            stdlib_bytes = sub_ch.encode('utf-16-le')
            print(f"U+{cp:05X}  {sub_ch}  →  high=U+{high:04X}  low=U+{low:04X}  "
                  f"bytes={stdlib_bytes.hex()}  restore=U+{restored:05X}")
print()


# ─────────────────────────────────────────────
# 4. BOM (Byte Order Mark)
# ─────────────────────────────────────────────

print("=== BOM ===\n")
s = "Привет"
for enc in ['utf-8-sig', 'utf-16', 'utf-32']:
    b = s.encode(enc)
    bom_hex = b[:4].hex()
    print(f"{enc:12s}: первые байты={bom_hex}  всего байт={len(b)}")

print("\nU+FEFF (BOM / Zero Width No-Break Space):")
bom = '\uFEFF'
print(f"  UTF-8:    {bom.encode('utf-8').hex()}")
print(f"  UTF-16LE: {bom.encode('utf-16-le').hex()}")
print(f"  UTF-16BE: {bom.encode('utf-16-be').hex()}")
print(f"  UTF-32LE: {bom.encode('utf-32-le').hex()}")
print()


# ─────────────────────────────────────────────
# 5. Длина строки: байты vs code points vs графемы
# ─────────────────────────────────────────────

import unicodedata

def grapheme_clusters(s: str) -> list[str]:
    """
    Простой алгоритм разбивки на графемные кластеры.
    Реальная реализация сложнее (UAX#29), но для ASCII и простых случаев работает.
    Для полной поддержки используйте: pip install grapheme
    """
    clusters = []
    current = ""
    for ch in s:
        ccc = unicodedata.combining(ch)
        if ccc == 0 and current:
            clusters.append(current)
            current = ch
        else:
            current += ch
    if current:
        clusters.append(current)
    return clusters


print("=== Длина строки на разных уровнях ===\n")
test_strings = [
    "Hello",
    "Привет",
    "🌍🎉",
    "é",          # U+00E9 (precomposed)
    "e\u0301",    # e + combining acute (decomposed)
    "👨‍👩‍👧",     # family emoji (ZWJ sequence)
    "🇷🇺",        # флаг России (regional indicator pair)
]

print(f"{'Строка':<20} {'bytes(utf8)':<13} {'code points':<13} {'graphemes':<10}")
print("-" * 60)
for s in test_strings:
    utf8_len = len(s.encode('utf-8'))
    cp_len   = len(s)  # в Python len() считает code points
    gr_len   = len(grapheme_clusters(s))
    repr_s = repr(s) if len(s) <= 5 else repr(s[:5]) + "…"
    print(f"{repr_s:<20} {utf8_len:<13} {cp_len:<13} {gr_len:<10}")
print()


# ─────────────────────────────────────────────
# 6. Ошибки при декодировании и стратегии обработки
# ─────────────────────────────────────────────

print("=== Стратегии обработки ошибок декодирования ===\n")
broken = b"Hello \xff\xfe World"  # невалидный UTF-8

strategies = ['strict', 'ignore', 'replace', 'backslashreplace', 'xmlcharrefreplace']
for strategy in strategies:
    try:
        result = broken.decode('utf-8', errors=strategy)
        print(f"  {strategy:20s}: {result!r}")
    except Exception as e:
        print(f"  {strategy:20s}: {type(e).__name__}: {e}")
print()


# ─────────────────────────────────────────────
# 7. Автоопределение кодировки
# ─────────────────────────────────────────────

print("=== Автоопределение кодировки (встроенные эвристики) ===\n")

samples = {
    "UTF-8 (русский)": "Привет мир".encode('utf-8'),
    "UTF-8 with BOM":  "Привет мир".encode('utf-8-sig'),
    "UTF-16-LE":       "Привет мир".encode('utf-16-le'),
    "UTF-16 with BOM": "Привет мир".encode('utf-16'),
    "ASCII":           b"Hello world",
    "Latin-1":         "Héllo".encode('latin-1'),
}

for name, data in samples.items():
    bom_hex = data[:4].hex()
    # Простая эвристика по BOM
    if data[:3] == b'\xef\xbb\xbf':
        detected = 'utf-8-sig'
    elif data[:2] == b'\xff\xfe':
        detected = 'utf-16-le'
    elif data[:2] == b'\xfe\xff':
        detected = 'utf-16-be'
    elif all(b < 0x80 for b in data):
        detected = 'ascii/utf-8'
    else:
        detected = '???  (нужен chardet/charset-normalizer)'
    print(f"  {name:25s}: BOM={bom_hex[:8]:8s}  →  {detected}")
print()
print("Для реального определения используйте: pip install charset-normalizer")

File: docs/03-encodings/examples.js
#!/usr/bin/env node
/**
 * Статья 3 — Кодировки: UTF-8, UTF-16, UTF-32.
 * Практические примеры на Node.js.
 *
 * Запуск: node examples.js
 */

// ─────────────────────────────────────────────
// 1. Кодирование строки в байты (через Buffer)
// ─────────────────────────────────────────────

console.log('=== Кодирование в байты (Buffer) ===\n');

function showEncodings(str) {
  console.log(`Строка: ${JSON.stringify(str)}`);
  for (const enc of ['utf8', 'utf16le', 'latin1', 'ascii']) {
    try {
      const buf = Buffer.from(str, enc);
      const hex = buf.toString('hex').replace(/(.{2})/g, '$1 ').trim();
      console.log(`  ${enc.padEnd(10)}: [${hex}]  (${buf.length} байт)`);
    } catch (e) {
      console.log(`  ${enc.padEnd(10)}: ошибка — ${e.message}`);
    }
  }
  // UTF-8 через TextEncoder (Web API, доступен в Node >= 11)
  const encoded = new TextEncoder().encode(str);
  const hex8 = [...encoded].map(b => b.toString(16).padStart(2, '0')).join(' ');
  console.log(`  TextEncoder: [${hex8}]  (${encoded.length} байт)`);
  console.log();
}

for (const ch of ['A', 'А', 'ñ', '🌍', '中']) {
  showEncodings(ch);
}


// ─────────────────────────────────────────────
// 2. Суррогатные пары: внутреннее представление JS
// ─────────────────────────────────────────────

console.log('=== Суррогатные пары в JS (UTF-16) ===\n');

function analyzeSurrogatePairs(str) {
  console.log(`Строка: ${JSON.stringify(str)}`);
  console.log(`  str.length (UTF-16 units): ${str.length}`);
  console.log(`  [...str].length (code points): ${[...str].length}`);
  for (let i = 0; i < str.length; i++) {
    const unit = str.charCodeAt(i);
    const hex = `0x${unit.toString(16).toUpperCase().padStart(4, '0')}`;
    let kind = '';
    if (unit >= 0xD800 && unit <= 0xDBFF) kind = ' ← high surrogate';
    if (unit >= 0xDC00 && unit <= 0xDFFF) kind = ' ← low surrogate';
    console.log(`  unit[${i}]: ${hex}${kind}`);
  }
  console.log();
}

for (const s of ['AB', '🌍', '𝄞', '🇷🇺']) {
  analyzeSurrogatePairs(s);
}


// ─────────────────────────────────────────────
// 3. Ручное вычисление суррогатной пары
// ─────────────────────────────────────────────

console.log('=== Суррогатные пары: вычисление вручную ===\n');

function toSurrogatePair(cp) {
  if (cp < 0x10000) throw new Error('BMP символ, суррогат не нужен');
  const cpPrime = cp - 0x10000;
  const high = 0xD800 + (cpPrime >> 10);
  const low  = 0xDC00 + (cpPrime & 0x3FF);
  return { high, low };
}

function fromSurrogatePair(high, low) {
  return 0x10000 + ((high - 0xD800) << 10) + (low - 0xDC00);
}

const smpChars = [0x1F30D, 0x1D11E, 0x20000, 0x1F3B5];
for (const cp of smpChars) {
  const { high, low } = toSurrogatePair(cp);
  const restored = fromSurrogatePair(high, low);
  const char = String.fromCodePoint(cp);
  console.log(
    `U+${cp.toString(16).toUpperCase().padStart(5, '0')}  ${char}  →` +
    `  high=U+${high.toString(16).toUpperCase()}` +
    `  low=U+${low.toString(16).toUpperCase()}` +
    `  restore=U+${restored.toString(16).toUpperCase().padStart(5, '0')}`
  );
}
console.log();


// ─────────────────────────────────────────────
// 4. TextEncoder / TextDecoder (Web API)
// ─────────────────────────────────────────────

console.log('=== TextEncoder / TextDecoder ===\n');

const encoder = new TextEncoder();  // всегда UTF-8
const decoder = new TextDecoder('utf-8');

const text = 'Hello, Мир! 🌍';
const bytes = encoder.encode(text);
const decoded = decoder.decode(bytes);

console.log(`Оригинал:    ${text}`);
console.log(`Байты (hex): ${Buffer.from(bytes).toString('hex')}`);
console.log(`Декодировано: ${decoded}`);
console.log(`Совпадает:   ${text === decoded}`);
console.log();

// Декодирование из UTF-16LE (Windows-стиль)
const utf16buf = Buffer.from('Привет', 'utf16le');
const decoded16 = new TextDecoder('utf-16le').decode(utf16buf);
console.log(`UTF-16LE bytes: ${utf16buf.toString('hex')}`);
console.log(`Decoded: ${decoded16}`);
console.log();


// ─────────────────────────────────────────────
// 5. Длина строки: code units vs code points vs graphemes
// ─────────────────────────────────────────────

console.log('=== Длина строки на разных уровнях ===\n');

function stringMetrics(s) {
  const utf8Len = Buffer.from(s, 'utf8').length;
  const utf16Units = s.length;
  const codePoints = [...s].length;
  let graphemes = 0;
  if (typeof Intl.Segmenter !== 'undefined') {
    const seg = new Intl.Segmenter('und', { granularity: 'grapheme' });
    graphemes = [...seg.segment(s)].length;
  } else {
    graphemes = '(Intl.Segmenter unavailable)';
  }
  return { utf8Len, utf16Units, codePoints, graphemes };
}

const tests = ['Hello', 'Привет', '🌍🎉', 'é', 'e\u0301', '👨‍👩‍👧', '🇷🇺'];
console.log(`${'строка'.padEnd(20)} ${'utf8'.padEnd(8)} ${'utf16'.padEnd(8)} ${'cp'.padEnd(8)} graphemes`);
console.log('─'.repeat(60));
for (const s of tests) {
  const { utf8Len, utf16Units, codePoints, graphemes } = stringMetrics(s);
  const repr = JSON.stringify(s).padEnd(20);
  console.log(`${repr} ${String(utf8Len).padEnd(8)} ${String(utf16Units).padEnd(8)} ${String(codePoints).padEnd(8)} ${graphemes}`);
}
console.log();


// ─────────────────────────────────────────────
// 6. Обнаружение BOM
// ─────────────────────────────────────────────

console.log('=== Обнаружение BOM ===\n');

function detectBOM(buf) {
  if (buf[0] === 0xEF && buf[1] === 0xBB && buf[2] === 0xBF) return 'UTF-8 BOM';
  if (buf[0] === 0xFF && buf[1] === 0xFE && buf[2] === 0x00 && buf[3] === 0x00) return 'UTF-32 LE BOM';
  if (buf[0] === 0x00 && buf[1] === 0x00 && buf[2] === 0xFE && buf[3] === 0xFF) return 'UTF-32 BE BOM';
  if (buf[0] === 0xFF && buf[1] === 0xFE) return 'UTF-16 LE BOM';
  if (buf[0] === 0xFE && buf[1] === 0xFF) return 'UTF-16 BE BOM';
  return 'No BOM';
}

const bomTests = [
  ['UTF-8 BOM',   Buffer.from([0xEF, 0xBB, 0xBF, 0x48, 0x65])],
  ['UTF-16 LE',   Buffer.from([0xFF, 0xFE, 0x48, 0x00])],
  ['UTF-16 BE',   Buffer.from([0xFE, 0xFF, 0x00, 0x48])],
  ['No BOM',      Buffer.from([0x48, 0x65, 0x6C, 0x6C])],
];

for (const [desc, buf] of bomTests) {
  console.log(`  ${desc.padEnd(15)}: BOM=${buf.subarray(0, 4).toString('hex').padEnd(10)}  → ${detectBOM(buf)}`);
}
console.log();


// ─────────────────────────────────────────────
// 7. Ловушки при работе с суррогатами
// ─────────────────────────────────────────────

console.log('=== Ловушки и граничные случаи ===\n');

// Нельзя slice по index — можно разрезать суррогатную пару
const emoji = '🌍World';
console.log(`'${emoji}'.slice(0, 2) = ${JSON.stringify(emoji.slice(0, 2))}  (НЕПРАВИЛЬНО — разрезает суррогат)`);
console.log(`[...'${emoji}'].slice(0, 2).join('') = ${[...emoji].slice(0, 2).join('')}  (правильно — через code points)`);
console.log();

// Regex без флага u может неправильно матчить
const emojiStr = '🌍🎉🚀';
console.log(`/${emojiStr}/ без флага u:`);
console.log(`  /./g matches: ${emojiStr.match(/./g)?.length}  (считает UTF-16 units!)`);
console.log(`  /./gu matches: ${emojiStr.match(/./gu)?.length}  (с флагом u — code points)`);
console.log();

// String.prototype.at() vs []  (Node >= 16)
if (emoji.at) {
  console.log(`emoji.at(0) = ${emoji.at(0)}  (правильно)`);
  console.log(`emoji[0] = ${emoji[0]}  (неправильно для суррогатов)`);
}
